// @mainframe/eth/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {
  ABI: true,
  BaseContract: true,
  ERC20: true,
  EthClient: true,
  Web3EthAbi: true,
  ETH_RPC_URLS: true
};
exports.ETH_RPC_URLS = exports.Web3EthAbi = exports.EthClient = exports.ERC20 = exports.BaseContract = exports.ABI = void 0;

var _abi = _interopRequireDefault(require("./abi"));

exports.ABI = _abi.default;

var _BaseContract = _interopRequireDefault(require("./Contracts/BaseContract"));

exports.BaseContract = _BaseContract.default;

var _ERC = _interopRequireDefault(require("./Contracts/ERC20"));

exports.ERC20 = _ERC.default;

var _Client = _interopRequireDefault(require("./Client"));

exports.EthClient = _Client.default;

var _web3EthAbi = _interopRequireDefault(require("./web3EthAbi"));

exports.Web3EthAbi = _web3EthAbi.default;

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _types[key];
});

var _utils = require("./utils");

Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _utils[key];
});
var ETH_RPC_URLS = {
  WS: {
    ropsten: 'wss://ropsten.infura.io/ws/v3/8ec0911ee74c4583b1346bbc1afdf22d',
    mainnet: 'wss://mainnet.infura.io/ws/v3/8ec0911ee74c4583b1346bbc1afdf22d',
    ganache: 'ws://localhost:8545'
  },
  HTTP: {
    ropsten: 'https://ropsten.infura.io/v3/8ec0911ee74c4583b1346bbc1afdf22d',
    mainnet: 'https://mainnet.infura.io/v3/8ec0911ee74c4583b1346bbc1afdf22d',
    ganache: 'http://localhost:8545'
  }
};
exports.ETH_RPC_URLS = ETH_RPC_URLS;
